### ISSUE TITLE
Add zookeeper configuration to runbook details

### ISSUE THREAD

**Issue by TylerJewell** [codenvy/codenvy#1369]
Discuss how to convert standard dockerized codenvy to zookeeper and the situations where this may be preferrable.

**Comment by bmicklea** [comment]
@skryzhny - can you please make some docs notes on how to configure Zookeeper with Codenvy?  It doesn't matter if we decide to use something else later, we want to capture how Zookeeper can be configured with Codenvy in the runbook.

**Comment by skryzhny** [comment]
There is 3 main steps:
1. Open in firewall on AIO port number 2181:
Edit file `/etc/puppet/modules/all_in_one/templates/iptables.erb` by inserting 2 lines: 

```
#etcd
-A INPUT -m state --state NEW -m tcp -p tcp --dport 2379 -j ACCEPT
...
#zookeeper
-A INPUT -m state --state NEW -m tcp -p tcp --dport 2181 -j ACCEPT
...
-A INPUT -m state --state NEW -m udp -p udp --dport 4789 -j ACCEPT
```

2. Create directories to hold zookeeper: `mkdir -p var/lib/zookeeper/datalog`
Start zookeep service by command:
`docker run --name codenvy-zookeeper --restart always -d -p 2181:2181 -v /var/lib/zookeeper/data:/data -v /var/lib/zookeeper/datalog:/datalog zookeeper`

3. Change configuration of docker daemon by editing file '/etc/puppet/modules/third_party/templates/docker/docker-network.erb' to looks like:
```
# /etc/sysconfig/docker-network
DOCKER_NETWORK_OPTIONS=' --bip=172.17.42.1/16 -H tcp://0.0.0.0:2375 -H unix:///var/run/docker.sock --cluster-store=zk://<%= scope.lookupvar('third_party::docker::install::docker_cluster_store') -%>:2181 --cluster-advertise=<%= scope.lookupvar('third_party::docker::install::docker_cluster_advertise') -%>:2375'
```
Basically it changes `--cluster-store` option.

Also I opened port 2181 at AWS EC2 firewall. That's all I did.

**Comment by TylerJewell** [comment]
@skryzhny  - I reopened this issue, because it is a docs issue, and we should not close it until we have captured everything and placed into runbook in the right way.

**Comment by bmicklea** [comment]
I already added it to the runbook

**Comment by TylerJewell** [comment]
@skryzhny - the configuration steps that you gave are for the puppetized version. Have you tried using ZooKeeper with the Dockerized version?  

I have some docs that we have published for setting up Codenvy to use Consul.  Could you do a similar sequence for ZooKeeper?  It's note quite clear to me what we should change in the instructions here:
https://github.com/codenvy/codenvy/blob/master/docs/_docs/admin-guide/admin-managing.md#scaling-with-overlay-network-linux-only

If you notice, there are 3 specific Docker daemon options that must be added to each workspace node after the key-value storage is configured.  Given your `docker run` command for ZooKeeper above, what would be the right syntax for those options?

**Comment by skryzhny** [comment]
Sorry, you really puzzled me.
How key-value storage is related to scaling?
And what is "workspace node" ?
And for me that doc is wrong and misleading.

**Comment by bmicklea** [comment]
You need KV storage to add nodes to swarm. A workspace node is a node which
the customer adds in order to run more workspaces so it's not part of the
original set of Codenvy Docker containers.

If the doc has wrong information please correct it. The section called
Pupput Based Version will be deleted so we need any of the information in
that section to be brought up to the Dockerized version section if it's
still useful. Or it may be easier to just write new content for the
dockerized version section.

The kind of questions I hear most are:
- What kind of storage can I use?
- Can the storage be NFS or does it have to be local?
- What kind of AWS servers are recommended?
- How do I setup the ports and network for AWS?

**Comment by skryzhny** [comment]
Sorry, You don't need to have KV storage to add nodes to swarm.
KV storage needed for overlay networking.
You may have several nodes in swarm and use them in bridged network mode without KV storage.

I think new term "Workspace Node" is really wrong and misleading.
We have term "MachineNode" what is basically equivalent to physical/virtual instance with docker installed.
"Workspace" can have several "Machines" that may run on different "MachineNodes".
So for me "Workspace Node" is some RUNTIME temporary thing, that belong to the user's workspace.

I didn't familiar with `consul`, I just noticed that doc about it is a strange mix of command that run on native linux and in boot2docker. I think it will not work, at least without starting several VM's before trying to run these commands.

**Comment by bmicklea** [comment]
Machine Nodes is fine you can use that term.

Although swarm doesn't require overlay networks our customers do since
they're all running compose files. So the config should assume an overlay
network and KV.

You are the expert on this stuff so change anything you think is wrong. It
should all work with Dockerized Codenvy on AWS (the config we'll run for
Codenvy.io).

**Comment by TylerJewell** [comment]
@skryzhny - please let me explain some concepts. 

It is true that you can use bridge networking to scale Codenvy. However, we are not going to allow customers to use this mode for two reasons:

1) There are very few customers that would ever get benefit from scaling swarm while using bridge mode because most customers will have multi-container deployments that require overlay networking. Our product would not be able to detect the differences between bridge and overlay networking well enough to prevent users from making mistakes on their workspace set up if the product was in bridge mode. So if a customer configured bridge mode and had weird behaviors, it would be hard to diagnose.  So it's better to just steer them towards the optimal solution.

2) @garagatyi essentially told me yesterday that even if we wanted to support scaling bridge mode with Codenvy, it wouldn't work. So that was that.

So, since overlay network is only required when you start to scale the product, we are discussing overlay networking in the scaling section of the managing guide. 

While it is technically accurate that these nodes are Machine Nodes, I do not want to use that term for our customers. To a customer admin, a workspace node is one where the workspaces are going to run. The master node is the one where Codenvy's master will run. The Codenvy master node and also be a workspace node.

About this action item - it's simpler than we think, so we don't have to make this too difficult.  From a product point of view, we are going to embed ZooKeeper runtime container into the Codenvy docker compose start. So the container will automatically be running.  But we haven't added this yet.  So, until that time, we just want to document how a customer can manually start ZooKeeper, configure various Docker daemons to work with it, and then modify the codenvy.env file to use it properly.

For the example on the managing.md file that is there, that is a Consul example that Alex prepared.  We have verified that it works.  We just want to have a similar example for ZooKeeper now.

I do not think anything belongs in the runbook, unless there are OS-specific attributes of ZooKeeper that we need to worry about, which we do not. 

So while you have done a puppetized configuration of ZooKeeper, I am just asking you to do the same for Dockerized version and show us any steps.

**Comment by TylerJewell** [comment]
Resolved as part of normal docs process.
